File: serial_client/unit_tests/reboot.py
#!/usr/bin/python3

# QMesh
# Copyright (C) 2021 Daniel R. Fay

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys, os
import pika
sys.path.insert(0, os.getcwd())
import qmesh_pb2
import qmesh_common

def dbg_process(ch, method, properties, body):
    ser_msg = qmesh_pb2.SerialMsg()
    ser_msg.ParseFromString(body)
    if(ser_msg.type == ser_msg.STATUS):
        qmesh_common.print_status_msg(ser_msg.status)
        if ser_msg.status.status == ser_msg.status.BOOTING:
            ch.stop_consuming()
            sys.exit(0)

# Set up the RabbitMQ connection
qmesh_common.setup_outgoing_rabbitmq(dbg_process)

# Reboot the board
qmesh_common.reboot_board()
qmesh_common.channel.start_consuming()

File: src/portability/reboot.cpp
/*
QMesh
Copyright (C) 2022 Daniel R. Fay

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "os_portability.hpp"
#include "serial_data.hpp"

volatile bool rebooting = false;

#if defined(MBED_OS)
void reboot_system() {
	rebooting = true;
    debug_printf(DBG_INFO, "Now rebooting the system...\r\n");
    portability::sleep(HALF_SECOND);
    NVIC_SystemReset();
}


#elif defined(ESP_IDF)
#include "components/esp_system/include/esp_system.h"
void reboot_system() {
	rebooting = true;
    debug_printf(DBG_INFO, "Now rebooting the system...\r\n");
    portability::sleep(HALF_SECOND);
    esp_restart();
}


#else
#error Need to define either MBED_OS or ESP_IDF

#endif